<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Bayes' Rule Practice Lab

_Author: Matt Brems_

This lab will have two parts:

- Part 1: Bayes Practice Problems
- Part 2: Moving from Probabilities to Distributions

---

## Part 1: Bayes Practice Problems

### Problem 1: IRAs

You work at a financial services company that offers individual retirement accounts (IRAs). To target high potential customers, you want to find out if people with children are more likely to invest in IRAs.  

So far you have found:
- 30% of all Americans have IRAs.
- half of Americans have children.
- two-thirds of those with IRAs have children.

Based on this information, what is the probability that someone with children has an IRA? 

In [10]:
P_A = 0.30  # Probability of having an IRA
P_B = 0.50  # Probability of having children
P_A_given_B = 2/3  # Probability of having an IRA given having children

# Probability of having children given having an IRA using Bayes theorem
P_B_given_A = (P_A_given_B * P_B) / P_A
P_B_given_A

1.1111111111111112

**Answer:** 1.1111

---

### Problem 2: Resident Satisfaction

Suppose I own a hotel and an apartment building. Times are tight, and I need to sell one of them. Either would earn me the exact same amount of money, so that doesn't help me to make my decision. Instead, I will make my decision based on my residents' satisfaction. 

- Whichever building has a higher percentage of satisfied people is the one I will choose to keep.
- *However*, if the percent of satisfied guests for my hotel and the percent of satified guests for my apartment building are within 10% of one another, I will sell neither until I can bring in some outside consultants to help me make my decision.
- Last winter, I administered a survey to *all* residents and found that 60% of them are satisfied.
- Among satisfied respondents, two out of every five came from the apartment building.
- One-third of my residents live in my apartment building.

What action will I take?

In [35]:
# Satisfaction percentages
overall_satisfaction = 60
apartment_satisfied_ratio = 2 / 5
apartment_residents_ratio = 1 / 3

apartment_satisfaction_percentage = apartment_satisfied_ratio * overall_satisfaction
overall_apartment_satisfaction = apartment_residents_ratio * apartment_satisfaction_percentage

# Check if the satisfaction percentages are within 10% of each other
threshold = 10
if abs(overall_apartment_satisfaction - overall_satisfaction) <= threshold:
    action = "Consult outside consultants"
else:
    action = "Sell the apartment building"

# action to take
action


'Sell the apartment building'

 **Answer:** 
 Sell the apartment building

### Problem 3: Coin Game

Suppose you and your friend are playing a game. Your friend has laid four coins out in front of you. If you flip heads, you win a dollar from your friend. If you flip tails, you owe a dollar to your friend. However, the coins in front of you are not fair.
- One coin has a 80% chance of flipping heads. (Call this coin A.)
- One coin has a 60% chance of flipping heads. (Call this coin B.)
- One coin has a 40% chance of flipping heads. (Call this coin C.)
- One coin has a 10% chance of flipping heads. (Call this coin D.)

**Problem 3 (a)**: Suppose you select one coin at random. That is, you don't know whether you selected coin A, B, C, or D. You flip heads. Given this data, what are the probabilities that you selected coin A, coin B, coin C, and coin D?

> Hint: When finding the marginal probability of flipping heads, you will need to use the Law of Total Probability! Find this **first**.

In [24]:
# Given probabilities
P_H_given_A = 0.80
P_H_given_B = 0.60
P_H_given_C = 0.40
P_H_given_D = 0.10

In [25]:
# Prior probabilities
P_A = 1/4
P_B = 1/4
P_C = 1/4
P_D = 1/4

In [26]:
# marginal probability of flipping heads
P_H = P_H_given_A * P_A + P_H_given_B * P_B + P_H_given_C * P_C + P_H_given_D * P_D

In [32]:
#Calculate the probabilities of selecting each coin given flipping heads
P_A_given_H = (P_H_given_A * P_A) / P_H
P_B_given_H = (P_H_given_B * P_B) / P_H
P_C_given_H = (P_H_given_C * P_C) / P_H
P_D_given_H = (P_H_given_D * P_D) / P_H


# Print the probabilities
print("Probability of selecting coin A given flipping heads:", P_A_given_H)
print("Probability of selecting coin B given flipping heads:", P_B_given_H)
print("Probability of selecting coin C given flipping heads:", P_C_given_H)
print("Probability of selecting coin D given flipping heads:", P_D_given_H)

Probability of selecting coin A given flipping heads: 0.4210526315789474
Probability of selecting coin B given flipping heads: 0.3157894736842105
Probability of selecting coin C given flipping heads: 0.2105263157894737
Probability of selecting coin D given flipping heads: 0.052631578947368425


**Answer**:
1. Probability of selecting coin A given flipping heads: 0.42
2. Probability of selecting coin B given flipping heads: 0.32
3. Probability of selecting coin C given flipping heads: 0.21
4. Probability of selecting coin D given flipping heads: 0.05

**BONUS: Problem 3 (b)**: Suppose you are using the same coin as before. That is, you _still_ don't know whether you selected coin A, B, C, or D - although you have a better idea now that you flipped heads on your first flip! On this second flip, you flip tails. Given this data, what are the probabilities that you selected coin A, coin B, coin C, and coin D?

> Hint: Everything now is **conditional on the fact that we have already flipped a heads**.

$$
\begin{eqnarray*}
P(A|B\cap C) &=& \frac{P(A\cap B \cap C)}{P(B\cap C)} \\
\\
&=& \frac{P(A\cap B|C)P(C)}{P(B|C)P(C)} \\
\\
&=& \frac{P(A\cap B|C)}{P(B|C)} \\
\end{eqnarray*}
$$

**Answer**:

**BONUS: Problem 3(c)**: Suppose you are using the same coin as before. That is, you _still_ don't know whether you selected coin A, B, C, or D - although you have a better idea now that you flipped heads, then tails on your first two flips! On this third flip, you flip tails. Given this data, what are the probabilities that you selected coin A, coin B, coin C, and coin D?

> Hint: This will follow the same structure as the last two problems.

**Answer**:

---

## Part 2: Moving from Probabilities to Distributions

### Problem 4: Prior Probabilities

**Problem 4 (a)**: In Problem 3, before you had flipped any coin, what was the probability that you had selected coin A? Coin B? Coin C? Coin D?

**Answer:** 
1. Probability of selecting coin A: P(A) = 1/4
2. Probability of selecting coin B: P(B) = 1/4
3. Probability of selecting coin C: P(C) = 1/4
4. Probability of selecting coin D: P(D) = 1/4



**Problem 4 (b)**: What is the definition of a distribution?
> Check your notes from the early weeks of the course if you can't remember.

**Answer:**

"Random variable with associated probabilities" #Hank Lesson 2.03


**Problem 4 (c)**: What "named distribution" (i.e. a distribution that has a common name) could we apply to a situation where selecting coins A-D is equally likely?

**Answer:** 

Discrete uniform distribution

**Problem 4 (d)**: Suppose that I have only one coin and I want to conduct inference on the probability of flipping heads, $p$. Note that $p$ is unknown. If I want to specify some prior distribution where all probabilities $p$ are equally likely, should I use a discrete or continuous distribution? Why?

**Answer:**

You should use a continuous distribution. Because "P" can take on any value between 0 and 1, and a continuous distribution allows for an infinite number of possible values within that range.

**Problem 4 (e)**: Suppose that I have only one coin and I want to conduct inference on the probability of flipping heads, $p$. Note that $p$ is unknown. If I want to specify some prior distribution where all probabilities $p$ are equally likely, then what named distribution might I use?

**Answer:**

You can use the Uniform distribution.

**Problem 4 (f)**: Suppose I have only one coin with some unknown probability of flipping heads $p$. If I think 50% is the likeliest value for $p$ and, as we get farther away from 50%, that value is less and less likely, what named distribution could I use?

**Answer:**

You can use the Beta distribution

**Problem 4 (g)**: When we listed out our prior probabilities for each coin, we got the same number of posterior probabilities. For example, when we had four prior probabilities (one for Coin A, one for Coin B, one for Coin C, and one for Coin D), then we had four posterior probabilities (one for Coin A, for Coin B, one for Coin C, and one for Coin D). However, if we have one prior distribution, then we will get one posterior distribution. What are some advantages to being able to summarize our posterior with one distribution instead of a big list of probabilities?

**Answer:**


Summarizing the posterior distribution with a single distribution instead of a list of individual probabilities has several advantages. It makes things simpler by giving a clear and concise picture of the uncertainty in the parameter we are interested in. Instead of dealing with a long list of probabilities, we have one distribution that represents all the possible values and how likely they are. This single distribution captures all the available information, including the data we have and any prior knowledge we have about the parameter. It makes calculations and analysis easier and more efficient. We can use the distribution to make decisions by considering the risks and rewards associated with different choices. When new data becomes available, updating the single distribution is straightforward, allowing us to keep refining our understanding of the parameter. Overall, summarizing with a single distribution simplifies things, gives us a complete picture, helps with decision-making, and allows us to easily update our knowledge as we learn more.